### PR TITLE
Dockerise portal

### PR DESCRIPTION
**Pull request by evanjt**
This PR creates a deployment of the Nomis data portal using docker. 

The versions are fixed using R's `remotes` library and the [Docker image from rocker](https://hub.docker.com/r/rocker/shiny).

### CODE REVIEW

**Comment by evanjt** [comment]
Hi @sbancal, I have containerised the application and would appreciate your input to review this PR. 

I have created an environment lockfile with the R library renv to fix library packages at the versions in the current running portal (nomis-data.epfl.ch), a Dockerfile to replicate the system setup as defined in https://github.com/ninojeannet/SBER-NOMIS-PORTAL/wiki/Deployment, a github workflow to auto build an image on PR/new version, and a docker-compose file for a local development environment.

If you wish to get this running locally, performing a `docker compose up --build` should be all that is needed on the application side, and pointing localhost to `nomis.local` in `/etc/hosts`.

You will notice there are some local mappings for some folders to manage assets:
```
      - ./data:/srv/shiny-server/data
      - ./protocols:/srv/shiny-server/www/protocols
      - ./db_backups:/srv/shiny-server/db_backups
      - ./log:/srv/shiny-server/log
```
I can supply some data to fill in the data and protocols folders if you run into any trouble, and also a recent database dump that will allow you to login.

File: Dockerfile [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+# Install necessary dependencies for the R repository
+RUN apt-get update && apt-get install -y \
+    curl lsb-release wget build-essential
+
+# Defining Zurich as local timezone
+ARG DEBIAN_FRONTEND=noninteractive
+ENV TZ="Europe/Zurich"
+RUN ln -fs /usr/share/zoneinfo/Europe/Zurich /etc/localtime
+RUN apt-get -y install tzdata
+RUN dpkg-reconfigure -f noninteractive tzdata
```
Comment (sbancal) [inline Dockerfile]: Globally, I'd encourage you to have a minimum of RUN steps, but rather chain all related commands into one RUN cmd && cmd && cmd ...

See https://stackoverflow.com/a/39330458/446302

That link speaks also about ordering of the commands (less moving parts first will enhance cache usage). Here, the order seems quite good as I see.

It speaks also about multi-staging. Have you investigate this ? I think we can gain much of disk space if applied.

File: Makefile [hunk 2]
```diff
@@ -0,0 +1,7 @@
+APPLICATION_NAME ?= nomis-data-portal
+ 
+build:
+	docker build --tag ${APPLICATION_NAME} .
+run:
+	docker compose up
```
Comment (sbancal) [inline Makefile]: I'd expect `-d` flag to be run as a daemon.

File: Makefile [hunk 3]
```diff
@@ -0,0 +1,7 @@
+APPLICATION_NAME ?= nomis-data-portal
+ 
+build:
+	docker build --tag ${APPLICATION_NAME} .
```
Comment (sbancal) [inline Makefile]: I'd say:
`docker compose build --pull --tag ${APPLICATION_NAME}`

What do you think ?

File: Makefile (hunk 2 above)
Comment (sbancal) [reply Makefile]: What about:
`docker compose up --remove-orphans --build -d`
?

File: README.md [hunk 4]
```diff
@@ -1,66 +1,78 @@
...
+## Getting started
+
+The service can be run locally or within a Docker environment. It requires a MySQL/MariaDB database to support it, and therefore requires the credentials of such a database to exist by supplying environment variables
+
+### Docker Compose
+
+Using the supplied `docker-compose.yml` file, a nomis portal, mariadb database and Traefik reverse proxy will start, simulating a complete deployment. From the root of the repository, and with Docker and Docker compose set-up, type:
+
+```bash
+docker compose up --build
```
Comment (sbancal) [inline README.md]: What about encourage to use the Makefile !?
`make run`

File: README.md [hunk 5]
```diff
@@ -1,66 +1,78 @@
...
+```
+
+You should be able to access the site at [http://nomis.local](http://nomis.local).
+
+### Docker (build and run)
+
+With the supplied `Dockerfile`, and from the root of the repository, build the image with:
+
+```bash
+docker build -t nomis-portal .
```
Comment (sbancal) [inline README.md]: What about encourage to use the Makefile !?
`make run`

File: docker-compose.yml [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+      - ./db_backups:/srv/shiny-server/db_backups
+      - ./log:/srv/shiny-server/log
+
+  nomis-db:
+    image: mariadb:10.2.41
+    environment:
+      MYSQL_ROOT_PASSWORD: nomisroot
+      MYSQL_DATABASE: nomis_db
+      MYSQL_USER: nomis_portal
+      MYSQL_PASSWORD: nomis
```
Comment (sbancal) [inline docker-compose.yml]: this also

File: docker-compose.yml [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+      - "traefik.enable=true"
+      - "traefik.http.routers.nomis_portal.rule=Host(`nomis.local`)"
+      - "traefik.http.services.nomis_portal.loadbalancer.server.port=3838"
+    environment:
+      NOMIS_ENV: prod
+      NOMIS_DB_HOSTNAME: nomis-db
+      NOMIS_DB_PORT: 3306
+      NOMIS_DB_USERNAME: nomis_portal
+      NOMIS_DB_PASSWORD: nomis
+      NOMIS_DB_NAME: nomis_db
```
Comment (sbancal) [inline docker-compose.yml]: I would inject such content from environment variables.
Usualy, we have a `.env` file for this.

File: app/start.sh [hunk 8]
```diff
@@ -0,0 +1,20 @@
...
+rm -f app_config.R
+
+# Add environment vars from env into R config. This is done
+# due to RShiny not passing system env vars to application
+printf "ENV <- '%s'\n" "$NOMIS_ENV" >> app_config.R
+printf "DB_NAME <- '%s'\n" "$NOMIS_DB_NAME" >> app_config.R
+printf "HOSTNAME <- '%s'\n" "$NOMIS_DB_HOSTNAME" >> app_config.R
+printf "DB_PORT <- %s\n" "$NOMIS_DB_PORT" >> app_config.R
+printf "USERNAME <- '%s'\n" "$NOMIS_DB_USERNAME" >> app_config.R
+printf "PASSWORD <- '%s'\n" "$NOMIS_DB_PASSWORD" >> app_config.R
```
Comment (sbancal) [inline app/start.sh]: What about simpler syntax:

```bash
cat <<EOF >app_config.R
ENV <- '$NOMIS_ENV'
DB_NAME <- '$NOMIS_DB_NAME'
HOSTNAME <- '$NOMIS_DB_HOSTNAME'
DB_PORT <- '$NOMIS_DB_PORT'
USERNAME <- '$NOMIS_DB_USERNAME'
PASSWORD <- '$NOMIS_DB_PASSWORD'
EOF
```

**Review by sbancal — COMMENTED** [review]
Hey @evanjt ,
Here are some suggestions regarding your PR.

Happy to talk about it if needed 😉

File: app/start.sh (hunk 8 above)
Comment (evanjt) [reply app/start.sh]: Agreed, this is a lot nicer

File: Dockerfile (hunk 1 above)
Comment (evanjt) [reply Dockerfile]: You're right, the steps set out were very explicit to copy 1:1 of the documented deployment steps. A functional deployment was prioritised over the final image size and efficiency.

How would you do a multi-stage Dockerfile in this case? Considering a lot of system libraries and specific versions are needed for the included R packages to function?

File: app/start.sh [hunk 9]
```diff
@@ -0,0 +1,22 @@
+#!/bin/bash
+# RShiny deployments do not pass system environment variables
+# to the application. This script grabs the env vars set in docker
+# and creates a config file which is read by the application on
+# initialisation, then starts the application
+
+# Delete any existing app config file
+rm -f app_config.R
```
Comment (sbancal) [inline app/start.sh]: not needed, as line 12 will overwrite the file anyway.

**Review by sbancal — APPROVED** [review]
LGTM, with tiny comment.

Do you want me to test it also ?

**Comment by evanjt** [comment]
> LGTM, with tiny comment.
> 
> Do you want me to test it also ?

No need, the Dockerfile is being built in github actions in the case you would like to validate the build steps.

**Review by sbancal — APPROVED** [review]
👍🏼

**Pull request merged**